src/output/plugins/visualization/VisualizationServer.hxx

```diff
@@ -0,0 +1,129 @@
+ */
+
+#ifndef VISUALIZATION_SERVER_HXX_INCLUDED
+#define VISUALIZATION_SERVER_HXX_INCLUDED 1
+
+#include "VisualizationClient.hxx"
+
+#include "SoundAnalysis.hxx"
+
+#include "config/Net.hxx"
+#include "event/CoarseTimerEvent.hxx"
+#include "event/ServerSocket.hxx"
+#include "net/SocketAddress.hxx"
+#include "net/UniqueSocketDescriptor.hxx"
+
+struct AudioFormat;
+struct ConfigBlock;
+class SoundInfoCache;
+
+/**
+ * \class VisualizationServer
+ *
+ * \brief A socker server handling visualization clients
+ *
+ * \sa \ref vis_out_arch "Architecture"
+ *
+ *
+ * This class handles binding one or more sockets & accepting incoming
+ * connections. For each such incoming connection, it will allocate a
+ * VisualizationClient instance to represent that client.
+ *
+ * The clients require both a PCM data cache and a SoundAnalysis instance to do
+ * their work. The former must be shared with the plugin that ultimately owns
+ * this class as well as the VisualizationClient instances, while the latter is
+ * cheaply copyable and so each client simply gets its own copy.
+ *
+ * The problem is that both must know the audio format in use (i.e. the number
+ * of samples per second and the number of channels), and that is only known
+ * when the plugin is "opened". Therefore this class can be represented by, yes,
+ * a finite state machine:
+ *
+ \code
+
+   Open --- OnPluginOpened() ---> HavePcmData
+     ^                                |
+     |                                |
+     +---- OnPluginClosed() ----------+
+
+ \endcode
+ *
+ * When a new client connection is opened:
+ *
+ * - if we are in state Open, we cannot provide the client with sound analysis
+ *   information nor a reference to the PCM cache
+ * - if we are in state HavePcmData, we can share a reference to our PCM cache
+ *   along with the salient information needed for sound analysis
+ *
+ * On state change:
+ *
+ * - from Open to HavePcmData, we can update all extant clients with a
+ *   shared reference to the PCM cache as well as the new sound analysis
+ *   information
+ * - from HavePcmData to Open, we need to tell all extant clients to
+ *   drop their PCM cache references, as well as their sound analysis
+ *   information
+ *
+ *
+ */
+
+class VisualizationServer : public ServerSocket {
+
+    /// only valid when the plugin is open
+	struct HavePcmData {
+        // I wish C++ had a `not_null` class
+        std::shared_ptr<SoundInfoCache> pcache;
+    };
+    /// Present state-- 👇 means closed, 👇 means opened (the plugin, that is)
+    std::variant<std::monostate, HavePcmData> state;
+	/// maximum number of clients permitted; zero => unlimited
+	size_t max_clients;
+
+	/// Clients have both a reference to the PCM cache as well as a
+	/// SoundAnalysis instance while the plugin is opened. We'll create new
+	/// clients with our present state
+	std::list<std::unique_ptr<VisualizationClient>> clients;
+	/// invoked periodically to clean-up dead clients
+	CoarseTimerEvent reaper;
```

Comment:
Why have a timer, why not remove dead clients right away?

Reply:
In order to remove dead clients right away, the server would need a means of notification. Other plugins use a "back reference" from client to server as this means. This, however, introduces a cyclic dependency between the two, which I prefer to avoid.
I wrote-up my thoughts [here](https://github.com/MusicPlayerDaemon/MPD/pull/1910/files#diff-90c68e20e2a42d1047afc0974fb97e354104c84a775ed9d3bbc039eae021b3f9R415).